enter.c

```diff
@@ -161,15 +161,55 @@ int enter(struct entry_settings *opts)
 		errx(1, "cannot create NICs when not in a network namespace");
 	}
 
+
+	char cgroup_path[BUFSIZ];
+	/*
+	 * If there is a cgroup specified we need to create bst.<pid> subcgroup to join
+	 * Otherwise if no cgroup is specified but limits are applied bst will error out
+	 */
+	if (opts->nactiveclimits != 0 || opts->cgroup_path != NULL) {
+		// If cgroup path is unspecified then attempt to derive from /proc/self/cgroup
+		if (opts->cgroup_path == NULL) {
+			FILE *selfcgroupfd = fopen("/proc/self/cgroup", "r");
+			if (selfcgroupfd == NULL) {
+				err(1, "unable to derive current cgroup hierarchy from /proc/self/cgroup");
+			}
+
+			char selfcgroup[BUFSIZ];
+
+			while (fgets(selfcgroup, BUFSIZ, selfcgroupfd) != NULL) {
+				if (strncmp(selfcgroup, "0::/", 3) == 0) {
+					// Remove newline character read by fgets
+					selfcgroup[strcspn(selfcgroup, "\n")] = '\0';
+					break;
+				}
+			}
+			fclose(selfcgroupfd);
+
+			// Form path from default cgroups path + current cgroup path with removed 0:/
+			if (sprintf(cgroup_path, "/sys/fs/cgroup/%s", selfcgroup + 3) == -1) {
```

Comment:
Adding 3 here feels wrong; also, sprintf is dangerous since it does not check the buffer size. It might be OK right now, since both buffers are the same size, but this is fragile as if the size gets changed in one of the buffers (which is the case for my PATH_MAX comment above), it'll just introduce at best a bug, at worst a security vulnerability.

How about something like this?

```c
const char *selfcgroup = NULL;
char line[BUFSIZ];
while (fgets(line, sizeof (line), selfcgroupfd) != NULL) {
	if (strncmp(line, "0::/", sizeof (line)) == 0) {
		// Remove newline character read by fgets
		line[strcspn(line, "\n")] = '\0';
		selfcgroup = line + 3;
		break;
	}
}
fclose(selfcgroupfd);

if (selfcgroup != NULL) {
	makepath(cgroup_path, "/sys/fs/cgroup/%s", selfcgroup);
}
```